HHH-19209 Verify and fix ID class generation for inner classes

cigaly opened the pull request:
Jira issue [HHH-19209](https://hibernate.atlassian.net/browse/HHH-19209)

- sorting record components when generating ID class
- added tests to check ID class generation from inner classes; fixed minor bug in generated class name
- test classes in core that are using entity (instead of ID) as parameter to `find` method are annotated with `@Exclude` to avoid failing tests

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------


[HHH-19209]: https://hibernate.atlassian.net/browse/HHH-19209?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

cigaly committed:
HHH-19209 Additional test cases
          - check generation of ID class for inner classes
          - check if record components of generated ID class are sorted by name

cigaly committed:
HHH-19209 Fixed name resolution for autogenerated ID class name when entity class is inner class

cigaly committed:
HHH-19209 Test classes that are using find by entity instead of ID are annotated with @Exclude

cigaly commented:
My assumption that record component must be alphabetically sorted by name was not correct, so All code and test based on the assumption are now removed.
Due to that change, three existing test cases are now failing:
`org.hibernate.orm.test.annotations.formula.JoinFormulaManyToOneLazyFetchingTest`,
`org.hibernate.orm.test.annotations.formula.JoinFormulaManyToOneNotIgnoreLazyFetchingTest`, and
`org.hibernate.orm.test.annotations.formula.JoinFormulaOneToOneNotIgnoreLazyFetchingTest`.
But this is not caused by changes in this PR.

**UPDATE**: (444 days later) At the end, it seems that my assumption was not so wrong, property names were sorted in some scenarios.

cigaly commented:
All tests are failing due to problem described in Jira issue [HHH-19214](https://hibernate.atlassian.net/browse/HHH-19214) [ see PR [#9824](https://github.com/hibernate/hibernate-orm/pull/9824) ]. This is not caused by changes made in this PR:

cigaly committed:
HHH-19209 Annotation test classes that are failing due to bug(s) unrelated to this issue with @Exclude

gavinking commented:
@cigaly Right now it is generating the id class with the name `Entity_.Id`, and I think I prefer that to the ugly `$`.

cigaly commented:
I am afraid that (like as with money) you can not avoid `$` with subclass. In `include` statement, you use `.`, but class name will always have `$` in the place of `.`. Like in ...

```
jshell> class X { class Y {}}
|  created class X

jshell> var c = X.Y.class
c ==> class X$Y
```

Something like split personality :-)

cigaly committed:
HHH-19209 Exclude test class from meta model generation - Envers is not supporting record as IdClass

gavinking commented:
Oh, my apologies, I did not read carefully enough and I thought we were in the annotation processor where the APIs shield us from the `$`. I guess I don't really understand what that code is doing.

cigaly commented:
There is also problem with Envers. It can not handle record as `IdClass` so I have to annotate one test class with `@Exclude`.

It seems not so difficult to fix that, but I am not sure if this can be done without changing some existing interfaces, so I would rather prefer not to do that on my own.

gavinking commented:
> There is also problem with Envers. It can not handle record as `IdClass`

Does this mean that id class generation breaks existing code that uses Envers?

cigaly commented:
Yes, under two conditions - if generated ID class is record and if Envers remains in pre-records age. Quickest (and, hopefully, temporary) fix can be to generate ID class as POJO, but the long term fix should to introduce Envers to record.

cigaly committed:
HHH-19209 Excluding class from meta model generation
          Problem is not connected with meta model generation, test is failing if ID class is manually added

gavinking commented:
Thanks so much for this @cigaly. Apologies that this took so long to merge. That was my fault.

gavinking commented:
Shit I should have squashed some of the commits here. Damn :-(
